base/iopause.3
.\" vim: tw=75
.TH iopause 3
.SH NAME
iopause \- stateful handling of events based on the poll interface or file descriptors

.SH SYNTAX
\fB#include \(dqioause.h\(dq\fR

int \fBiopause\fP(iopause_fd *\fIx\fR, unsigned int \fIlen\fR, struct taia *\fIdeadline\fR, struct taia *\fIstamp\fR);

.SH DESCRIPTION
\fBiopause\fR checks for file descriptor readability or writeability as
specified by \fIx\fR[0].fd, \fIx\fR[0].events, \fIx\fR[1].fd,
\fIx\fR[1].events, ..., \fIx\fR[\fIlen\fR-1].fd,
\fIx\fR[\fIlen\fR-1].events. If \fIx\fR[i].events includes the bit
IOPAUSE_READ, \fBiopause\fR checks for readability of the descriptor
\fIx\fR[i].fd; if \fIx\fR[i].events includes the bit IOPAUSE_WRITE,
\fBiopause\fR checks for writability of the descriptor \fIx\fR[i].fd; other
bits in \fIx\fR[i].events have undefined effects.

\fBiopause\fR sets the IOPAUSE_READ bit in \fIx\fR[i].revents if it finds
that \fIx\fR[i].fd is readabled and it sets the IOPAUSE_WRITE bit in
\fIx\fR[i].revents if it finds that \fIx\fR[i].fd is writable. Beware that
readability and writeability may be destroyed at any moment by other
processes with access to the same file \fIx\fR[i].fd refers to.

If there is no readability or writeability to report, \fBiopause\fR waits
until \fIdeadline\fR for something to happen.  \fBiopause\fR will return
before \fIdeadline\fR if a descriptor becomes readable or writeable, or an
interrupting signal arrives, or some system-defined amount of time passes.
\fBiopause\fR sets \fIrevents\fR in any case.

You must put a current timestamp into \fIstamp\fR before calling
\fBiopause\fR. In case the the current timestamp is older than the previous
one (ie. due to daylight-saving) negative values are omitted.

.SH "IMPLEMENTATION NOTES"
The current implementation of \fBiopause\fR uses the \fBpoll\fR interface
if that is available.  On some systems, \fBpoll\fR needs to dynamically
allocate kernel memory. In case not enough memory is available,
\fBiopause\fR will return immediately and will report (often incorrectly)
that no descriptors are readable or writeable.

If the \fBpoll\fR interface is not available, \fBiopause\fR uses the
\fBselect\fR function. This function cannot see descriptor numbers past a
system-defined limit, typically 256 or 1024; \fBiopause\fR will
artificially pretend that those descriptors are never readable or
writeable.

Future implementations of \fBiopause\fR may work around these problems on
some systems, at the expense of chewing up all available CPU time.

Both \fBpoll\fR and \fBselect\fR use relative timeouts rather than absolute deadlines.
Some kernels round the timeout down to a multiple of 10 milliseconds; this
can burn quite a bit of CPU time as the deadline approaches. \fBiopause\fR
compensates for this by adding 20 milliseconds to the timeout.  In case the
current timestamp is older than the previous one (ie. due to
daylight-saving) the timeout is set to 20 milliseconds.

.SH "RETURN CODES"
\fBiopause\fR reads and deploys the return codes of the \fIpoll\fR and
\fIselect\fR call. Only positive return values shall be considered by the
calling routines for a succcessful invocation.

.SH CREDITS
Parts of the description are taken from the 'libowfat' man page.

.SH "SEE ALSO"
select(2),
poll(3),
taia(3)


ChangeLog
* ChangeLog
Release 1.0 Start 21/05/2020
01. added --release option to specify release on command line
- 13/09/2020
02. libqmail.spec: added Provides: qmail-libs
- 15/09/2020
03. added mess822, commands, rewritehost, sconfig functions from indimail-mta
- 16/09/2020
04. FreeBSD fixes
- 01/10/2020
05. Darwin Port (fmt_ushort.c, default.configure, configure.ac)
- 04/10/2020
06. fixed broken crypt(3) implemetation on Darwin Mac OS X
07. reorgnized directory structure. All qmail functions moved to 'base'
08. textcode/fmt_utf8.c: updated with upstream libowfat-0.32
- 09/10/2020
09. replaced fmt_tohex() with tohex()
10. global.h, hmac_md5.c, hmac_ripemd.c, md5.c, md5.h: replaced updated
    functions from indimail-mta
- 18/10/2020
11. alloc_re.c: fix overwriting unallocate memory location when
    new_size < old_size


base/iopause.c
/*
 * $Log: iopause.c,v $
 * Revision 1.2  2020-08-03 17:58:55+05:30  Cprogrammer
 * fixed formatting
 *
 * Revision 1.1  2003-12-31 19:46:55+05:30  Cprogrammer
 * Initial revision
 *
 */
#include "taia.h"
#include "select.h"
#include "iopause.h"

void
iopause(iopause_fd *x, unsigned int len, struct taia *deadline, struct taia *stamp)
{
	struct taia     t;
	int             millisecs;
	double          d;
	int             i;

	if (taia_less(deadline, stamp))
		millisecs = 0;
	else {
		t = *stamp;
		taia_sub(&t, deadline, &t);
		if ((d = taia_approx(&t)) > 1000.0)
			d = 1000.0;
		millisecs = d * 1000.0 + 20.0;
	}
	for (i = 0; i < len; ++i)
		x[i].revents = 0;

#ifdef IOPAUSE_POLL

	poll(x, len, millisecs);
	/*- XXX: some kernels apparently need x[0] even if len is 0 */
	/*- XXX: how to handle EAGAIN? are kernels really this dumb?  */
	/*- XXX: how to handle EINVAL? when exactly can this happen?  */

#else
	{

		struct timeval  tv;
		fd_set          rfds;
		fd_set          wfds;
		int             nfds;
		int             fd;

		FD_ZERO(&rfds);
		FD_ZERO(&wfds);

		nfds = 1;
		for (i = 0; i < len; ++i) {
			fd = x[i].fd;
			if (fd < 0)
				continue;
			if (fd >= 8 * sizeof(fd_set))
				continue; /*XXX*/
			if (fd >= nfds)
				nfds = fd + 1;
			if (x[i].events & IOPAUSE_READ)
				FD_SET(fd, &rfds);
			if (x[i].events & IOPAUSE_WRITE)
				FD_SET(fd, &wfds);
		}

		tv.tv_sec = millisecs / 1000;
		tv.tv_usec = 1000 * (millisecs % 1000);

		if (select(nfds, &rfds, &wfds, (fd_set *) 0, &tv) <= 0)
			return;
		/*- XXX: for EBADF, could seek out and destroy the bad descriptor */

		for (i = 0; i < len; ++i) {
			if ((fd = x[i].fd) < 0)
				continue;
			if (fd >= 8 * sizeof(fd_set))
				continue; /*XXX*/
			if (x[i].events & IOPAUSE_READ && FD_ISSET(fd, &rfds))
				x[i].revents |= IOPAUSE_READ;
			if (x[i].events & IOPAUSE_WRITE && FD_ISSET(fd, &wfds))
				x[i].revents |= IOPAUSE_WRITE;
		}

	}
#endif

}
